### PR TITLE
gh-90848: Fixed create_autospec ignoring configure_mock style kwargs

### PR DESCRIPTION
**Pull request by infohash**
# Fix for:

* Issue: gh-90848

# Description
All mocks can be pre-configured with `side_effect` and `return_value` by passing them to `kwargs` which are consumed by the [Mock class](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock). But `unittest.mock.create_autospec` function is ignoring kwargs so, the user has to configure them later in the code after the initialization of the mock.

# Test To Reproduce
```python
from unittest import mock

import pytest


class Result:
    def get_result(self):
        pass


@pytest.mark.xfail
def test_result():
    class_mock = create_autospec(spec=Result, **{'return_value.get_result.side_effect': ValueError})
    with pytest.raises(ValueError):
        classmock().get_result()


def test_result_configure_later():
    class_mock = create_autospec(spec=Result)
    class_mock.return_value.get_result.side_effect = ValueError
    with pytest.raises(ValueError):
        classmock().get_result()
```

# Details
This is the line 👇 which is consuming `kwargs` during the initialization of the mock:
https://github.com/python/cpython/blob/85f727c5fb2afa60affa9ae3396ce4149cf5215d/Lib/unittest/mock.py#L2774-L2777

Here 👇 we are overriding the `return_value` of the mock by creating a new mock on it but this is not the problem:
https://github.com/python/cpython/blob/85f727c5fb2afa60affa9ae3396ce4149cf5215d/Lib/unittest/mock.py#L2793-L2797

As it is being called recursively 👆, the attributes of the _specced_ object will also be mocked by creating **child mocks** 👇:
https://github.com/python/cpython/blob/85f727c5fb2afa60affa9ae3396ce4149cf5215d/Lib/unittest/mock.py#L2838-L2841

Then the child mocks are set on the recursive mock 👇:
https://github.com/python/cpython/blob/85f727c5fb2afa60affa9ae3396ce4149cf5215d/Lib/unittest/mock.py#L2849-L2851

This 👆 is the cause of the issue. We have set new child attribute mock which is not configured.

# Fix Suggested by @tirkarthi 

```diff

+    if _is_instance_mock(mock) and kwargs:
+        mock.configure_mock(**kwargs)

      return mock
```

# Code Changes
1. Added the above patch which will configure the parent mock **only** at the end because the user will pass `kwargs` with respect to the parent mock.
2. Renamed local variable `kwargs` to `child_kwargs` because it was shadowing the name of the argument which is now being used to configure the parent mock.
3. This 👇 should be popped because it will not be used again when configuring the parent mock at the end.
```diff
-    wrapped = kwargs.get('wraps')
+    # Pop wraps from kwargs because it must not be passed to configure_mock.
+    wrapped = kwargs.pop('wraps', None)
```

### CODE REVIEW

File: Lib/test/test_unittest/testmock/testmock.py [hunk 1]
```diff
@@ -115,6 +115,26 @@ def f(): pass
...
+        class Result:
+            def get_result(self):
+                pass
+
+        mocked_result = 'mocked value'
+        class_mock = create_autospec(spec=Result, **{
+            # Test both side_effect & return_value which is why the 2nd
+            # parameter of side_effect is DEFAULT.
+            'return_value.get_result.side_effect': [ValueError, DEFAULT],
+            'return_value.get_result.return_value': mocked_result})
```
Comment (cjw296) [inline Lib/test/test_unittest/testmock/testmock.py]: Where else is this style of "string with docs indicate traverse through attributes" used?

File: Lib/test/test_unittest/testmock/testmock.py [hunk 2]
```diff
@@ -115,6 +115,26 @@ def f(): pass
         with self.assertRaises(TypeError):
             mock()
 
+    def test_create_autospec_should_be_configurable_by_kwargs(self):
+        """If kwargs are given to set side_effect & return_value, the function
+        must configure the parent mock during initialization."""
+        class Result:
+            def get_result(self):
```
Comment (cjw296) [inline Lib/test/test_unittest/testmock/testmock.py]: Please make sure this method is called at some point or add a `# pragma: no cover` to the end.
(code coverage is checked in the backport)

File: Lib/unittest/mock.py [hunk 3]
```diff
@@ -2788,8 +2788,8 @@ def create_autospec(spec, spec_set=False, instance=False, _parent=None,
     if _parent is not None and not instance:
         _parent._mock_children[_name] = mock
 
-    wrapped = kwargs.get('wraps')
-
+    # Pop "wraps" from kwargs because it will not be used again.
+    wrapped = kwargs.pop('wraps', None)
```
Comment (cjw296) [inline Lib/unittest/mock.py:2792]: Okay, but what harm does it do being there?
What happens if you revert this change?

Comment (infohash) [reply Lib/unittest/mock.py:2792]: https://github.com/python/cpython/blob/7c4107be8f482f97eee3086a53c9d06e59a64b92/Lib/unittest/mock.py#L2794-L2796

When `wraps` is given and if it is not popped from `kwargs`, it will make this condition 👇 true when `create_autospec` is called rescursively.

https://github.com/python/cpython/blob/7c4107be8f482f97eee3086a53c9d06e59a64b92/Lib/unittest/mock.py#L2853-L2854

It should be true only for the parent mock because the user will pass configure mock kwargs with respect to the parent attribute.

File: Lib/test/test_unittest/testmock/testmock.py (hunk 1 above)
Comment (infohash) [reply Lib/test/test_unittest/testmock/testmock.py]: Removed noisy comments.

File: Lib/test/test_unittest/testmock/testmock.py (hunk 2 above)
Comment (infohash) [reply Lib/test/test_unittest/testmock/testmock.py]: Removed the class and reusing the class which is already defined at top of the module.

**Review by cjw296 — APPROVED** [review]
